src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -66,84 +54,80 @@ module Make (Inputs : Intf.Main_inputs) = struct
+  (* Without [is_accepting_diffs], we would not be able to accept work
+     from the buffer if the worker_writer pipe is closed *)
+  let close_and_finish_copy t =
+    t.is_accepting_diffs <- false ;
+    let%bind () = flush t in
+    Strict_pipe.Writer.close t.worker_writer ;
+    let%map () = t.worker_thread in
+    Worker.close t.worker
 
-  let close_and_finish_copy {worker; stop_signal; worker_thread; _} =
-    Ivar.fill stop_signal `Stop ;
-    let%map () = worker_thread in
-    Worker.close worker
+  let select_work ({max_buffer_capacity; flush_capacity; buffer; _} as t)
+      current_work =
+    if
+      Queue.length buffer >= flush_capacity
+      && Deferred.is_determined current_work
+    then flush t
+    else if Queue.length buffer > max_buffer_capacity then
+      Debug_assert.debug_assert_deferred
+      @@ fun () ->
+      failwithf
+        !"There is too many work that a Transition Frontier Persistence \
+          worker is waiting for. Retune buffer parameters: {flush_capacity: \
+          %i, buffer_capacity: %i}"
+        flush_capacity max_buffer_capacity ()
+    else current_work
 
   let listen_to_frontier_broadcast_pipe
       (frontier_broadcast_pipe :
-        Transition_frontier.t option Broadcast_pipe.Reader.t)
-      ({max_buffer_capacity; flush_capacity; buffer; stop_signal; _} as t) =
+        Transition_frontier.t option Broadcast_pipe.Reader.t) t =
     Broadcast_pipe.Reader.iter frontier_broadcast_pipe
       ~f:
         (Option.value_map ~default:Deferred.unit ~f:(fun frontier ->
-             Broadcast_pipe.Reader.iter
-               (Transition_frontier.persistence_diff_pipe frontier)
-               ~f:(fun new_diffs ->
-                 let enqueue_and_flush_work () =
-                   Queue.enqueue_all buffer new_diffs ;
-                   if Queue.length buffer >= max_buffer_capacity then
-                     (* We would only have failure when we are debugging code  *)
-                     Debug_assert.debug_assert_deferred
-                     @@ fun () ->
-                     failwithf
-                       !"There is too many work that a Transition Frontier \
-                         Persistence worker is waiting for. Retune buffer \
-                         parameters: {flush_capacity: %i, buffer_capacity: %i}"
-                       flush_capacity max_buffer_capacity ()
-                   else if Queue.length buffer >= flush_capacity then flush t
-                   else Deferred.unit
-                 in
-                 if Ivar.is_empty stop_signal then enqueue_and_flush_work ()
-                 else Deferred.unit ) ))
+             Deferred.join
+             @@ Broadcast_pipe.Reader.fold ~init:Deferred.unit
+                  (Transition_frontier.persistence_diff_pipe frontier)
+                  ~f:(fun worker_thread new_diffs ->
+                    Deferred.return
+                    @@
+                    if t.is_accepting_diffs then (
```

Comment:
I'm not sure if I follow why this variable is necessary. shouldn't you just be able to close this pipe? If someone sets this to false from the outside, you will have info in the buffer that you never flush and process. Even if the code that sets this to false calls flush, that doesn't mean there isn't still information that will be received by this pipe handler afterwards. Closing the pipe will let the existing work finish and you can bind on the deferred from the pipe read in order to guarantee a final flush after all the data in the pipe has been processed.